luigi/contrib/beam_dataflow.py

```diff
@@ -0,0 +1,451 @@
+            "InProcessPipelineRunner",
+            "BlockingDataflowPipelineRunner"
+        ]:
+            logger.warning("Using deprecated runner %s. Consider upgrading to "
+                           "Beam 2.x." % self.runner)
+            return self.runner
+
+        else:
+            logger.warning("Found unsupported runner %s. Defaulting to "
+                           "DirectRunner." % self.runner)
+            return "DirectRunner"
+
+    def _get_dataflow_args(self):
+        def f(key, value):
+            return '--{}={}'.format(key, value)
+
+        output = []
+
+        runner = self._get_runner()
+        if runner:
+            output.append(f(self.dataflow_params.runner, runner))
+        if self.project:
+            output.append(f(self.dataflow_params.project, self.project))
+        if self.zone:
+            output.append(f(self.dataflow_params.zone, self.zone))
+        if self.region:
+            output.append(f(self.dataflow_params.region, self.region))
+        if self.staging_location:
+            output.append(f(self.dataflow_params.staging_location, self.staging_location))
+        if self.temp_location:
+            output.append(f(self.dataflow_params.temp_location, self.temp_location))
+        if self.gcp_temp_location:
+            output.append(f(self.dataflow_params.gcp_temp_location, self.gcp_temp_location))
+        if self.num_workers:
+            output.append(f(self.dataflow_params.num_workers, self.num_workers))
+        if self.autoscaling_algorithm:
+            output.append(f(self.dataflow_params.autoscaling_algorithm, self.autoscaling_algorithm))
+        if self.max_num_workers:
+            output.append(f(self.dataflow_params.max_num_workers, self.max_num_workers))
+        if self.disk_size_gb:
+            output.append(f(self.dataflow_params.disk_size_gb, self.disk_size_gb))
+        if self.worker_machine_type:
+            output.append(f(self.dataflow_params.worker_machine_type, self.worker_machine_type))
+        if self.worker_disk_type:
+            output.append(f(self.dataflow_params.worker_disk_type, self.worker_disk_type))
+        if self.network:
+            output.append(f(self.dataflow_params.network, self.network))
+        if self.subnetwork:
+            output.append(f(self.dataflow_params.subnetwork, self.subnetwork))
+        if self.job_name:
+            output.append(f(self.dataflow_params.job_name, self.job_name))
+        if self.service_account:
+            output.append(f(self.dataflow_params.service_account, self.service_account))
+        if self.labels:
+            output.append(f(self.dataflow_params.labels, json.dumps(self.labels)))
+
+        return output
+
+    def _format_input_args(self):
+        job_input = self.input()
+        if isinstance(job_input, luigi.Target):
+            job_input = {"input": job_input}
+        elif not isinstance(job_input, dict):
+            raise ValueError("Input (requires()) must be dict type")
```

Comment:
thanks! it looks like [input()](https://github.com/spotify/luigi/blob/62fbaf20775b69995cacb3e266a8b50a6959fd92/luigi/task.py#L636) does convert Tasks to a `list` of their respective paths, so I definitely ned to add that. do you know where `tuple` output support is implemented?

Reply:
also -- do you have any thoughts on how a flat `list` of Targets should get represented? for example if our `requires()` is a `Task` like this: 

```
class TestRequires(luigi.ExternalTask):
    def output(self):
        return [
          luigi.LocalTarget(path='some-input-dir'),
          luigi.LocalTarget(path='some-input-dir2')
        ]
```

Should that get formatted as something like `input1=some-input-dir, input2=some-input-dir2`, or `input=some-input-dir1,some-input-dir2`?

Reply:
hey @honnix just wanted to ping this discussion, looks like it's the main outstanding issue on the PR?

Reply:
Sorry for replying late. We will take a look at this.

Reply:
`tuple` would fall into this branch: https://github.com/spotify/luigi/blob/master/luigi/task.py#L883

Regarding how to represent `input`, I'm not very familiar with Dataflow options. I think either way if fine as long as it can be interpreted correctly.

Reply:
cool, I updated it to support a list of `Target`s, a list of `(key, Target)` tuples, and a single tuple. The `Task` type is handled in the background in the [getpaths()](https://github.com/spotify/luigi/blob/master/luigi/task.py#L838) method. For a flat list of `Targets` I chose to represent it as `--input=some-input-dir1,some-input-dir2`.

Reply:
I think now this error message is not accurate anymore. How about just `"Invalid input (requires())"? Or we could give a list of supported options in the message.

Reply:
thanks for catching this, fixed.